pkg/rc/replication_controller.go

```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	} else {
-		modifyIndex = queryMeta.LastIndex
-	}
+	newNode := newNodes[0]
 
-	status.NodeTransfer = &rcstatus.NodeTransfer{
-		// Again, we only support one node transfer at a time so just take the first old and new
-		OldNode: ineligibleCurrent[0],
-		NewNode: newNodes[0],
+	status := rcstatus.Status{
+		NodeTransfer: &rcstatus.NodeTransfer{
+			OldNode: oldNode,
+			NewNode: newNode,
+		},
 	}
+
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
 	if err != nil {
-		return util.Errorf("Could not write new node to store: %s", err)
+		return "", util.Errorf("Could not write new node to store: %s", err)
 	}
 
 	err = transaction.MustCommit(writeCtx, rc.txner)
 	if err != nil {
-		return util.Errorf("Could not commit CASTxn: %s", err)
+		return "", util.Errorf("Could not commit CASTxn: %s", err)
+	}
+
+	return newNode, nil
+}
+
+func (rc *replicationController) scheduleWithSession(newNode types.NodeName) error {
+	rc.logger.NoFields().Infof("Scheduling on %s", newNode)
+	rc.mu.Lock()
+	manifest := rc.Manifest
+	rc.mu.Unlock()
+
+	manifestBytes, err := manifest.Marshal()
+	if err != nil {
+		return err
+	}
+
+	key := path.Join(string(consul.INTENT_TREE), newNode.String(), manifest.ID().String())
+	name := fmt.Sprintf("wait-on-%s-health", newNode)
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	defer func() {
+		// We close the context on error to prevent a leak but leave it
+		// otherwise so it can be used to cancel the session. We write the
+		// defer this way so that reassignments to err are noticed and the
+		// final value is used.
+		if err != nil {
+			ctxCancel()
+		}
+	}()
+
+	_, session, err := consul.SessionContext(ctx, rc.consulClient, name)
+	if err != nil {
+		return err
+	}
+
+	rc.nodeTransfer.session = session
+
+	// Call this function to clean up the session when the node
+	// transfer finishes or is stopped
+	rc.nodeTransfer.sessionCancel = ctxCancel
+
+	writeCtx, writeCancel := transaction.New(context.Background())
+	defer writeCancel()
+
+	txnUnlocker, err := session.LockIfKeyNotExistsTxn(writeCtx, key, manifestBytes)
+	if err != nil {
+		return nil
+	}
+	rc.nodeTransfer.unlocker = txnUnlocker
+
+	ok, resp, err := transaction.Commit(writeCtx, rc.txner)
+	switch {
+	case err != nil:
+		return err
+	case !ok:
+		return util.Errorf("could not schedule new node due to a transaction violation: %s", transaction.TxnErrorsToString(resp.Errors))
 	}
 
 	return nil
 }
 
-func (rc *replicationController) transferPetNodes(ineligibleCurrent []types.NodeName) error {
-	errMsg := fmt.Sprintf("RC has scheduled %d ineligible nodes: %s", len(ineligibleCurrent), ineligibleCurrent)
-	err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+func (rc *replicationController) watchHealth() {
+	defer func() {
+		rc.isWatchingHealth = false
```

Comment:
you're going to have a data race on `isWatchingHealth` because multiple goroutines write to it. You could guard it with a mutex to fix that